src/node/oscillator.rs

```diff
@@ -0,0 +1,1238 @@
+#[allow(clippy::pedantic)]
+pub struct OscillatorOptions {
+    pub type_: Option<OscillatorType>,
+    pub frequency: Option<f32>,
+    pub detune: Option<f32>,
+    pub channel_config: Option<ChannelConfigOptions>,
+    pub periodic_wave: Option<PeriodicWave>,
+}
+
+impl Default for OscillatorOptions {
+    fn default() -> Self {
+        Self {
+            type_: Some(OscillatorType::default()),
+            frequency: Some(440.),
+            detune: Some(0.),
+            channel_config: Some(ChannelConfigOptions::default()),
+            periodic_wave: None,
+        }
+    }
+}
+
+/// Waveform of an oscillator
+#[derive(Debug, Copy, Clone, PartialEq)]
+// the naming comes from the web audio specfication
+#[allow(clippy::pedantic)]
+pub enum OscillatorType {
+    Sine,
+    Square,
+    Sawtooth,
+    Triangle,
+    Custom,
+}
+
+impl Default for OscillatorType {
+    fn default() -> Self {
+        Self::Sine
+    }
+}
+
+impl From<u32> for OscillatorType {
+    fn from(i: u32) -> Self {
+        use OscillatorType::{Custom, Sawtooth, Sine, Square, Triangle};
+
+        match i {
+            0 => Sine,
+            1 => Square,
+            2 => Sawtooth,
+            3 => Triangle,
+            4 => Custom,
+            _ => unreachable!(),
+        }
+    }
+}
+
+/// Message types used to communicate between [`OscillatorNode`] and [`OscillatorRenderer`]
+enum OscMsg {
+    PeriodicWaveMsg {
+        computed_freq: f32,
+        wavetable: Vec<f32>,
+        norm_factor: f32,
+        disable_normalization: bool,
+    },
+}
+
+/// Audio source generating a periodic waveform
+// the naming comes from the web audio specfication
+#[allow(clippy::pedantic)]
+pub struct OscillatorNode {
+    registration: AudioContextRegistration,
+    channel_config: ChannelConfig,
+    sample_rate: f32,
+    frequency: AudioParam,
+    detune: AudioParam,
+    type_: Arc<AtomicU32>,
+    scheduler: Scheduler,
+    sender: Sender<OscMsg>,
+}
+
+impl AudioScheduledSourceNode for OscillatorNode {
+    fn scheduler(&self) -> &Scheduler {
+        &self.scheduler
+    }
+}
+
+impl AudioNode for OscillatorNode {
+    fn registration(&self) -> &AudioContextRegistration {
+        &self.registration
+    }
+
+    fn channel_config_raw(&self) -> &ChannelConfig {
+        &self.channel_config
+    }
+
+    /// `OscillatorNode` is a source node. A source node is by definition with no input
+    fn number_of_inputs(&self) -> u32 {
+        0
+    }
+
+    /// `OscillatorNode` is a mono source node.
+    fn number_of_outputs(&self) -> u32 {
+        1
+    }
+}
+
+impl OscillatorNode {
+    /// Returns an `OscillatorNode`
+    ///
+    /// # Arguments:
+    ///
+    /// * `context` - The `AudioContext`
+    /// * `options` - The Oscillatoroptions
+    pub fn new<C: AsBaseAudioContext>(context: &C, options: Option<OscillatorOptions>) -> Self {
+        context.base().register(move |registration| {
+            let sample_rate = context.base().sample_rate().0 as f32;
```

Comment:
Would it be possible to replace the SampleRate(u32) by an f32 ?
It would:
- avoid castings (with possible loss of precision)
- be compliant with the [specs](https://www.w3.org/TR/webaudio/#dom-baseaudiocontext-samplerate)